blog/_posts/2016-02-03-essence-of-scala.md

```diff
@@ -0,0 +1,153 @@
+studied formally. Translated to Scala notation, the language covered
+by DOT is described by the following abstract grammar:
+
+    Value       v  =  (x: T) => t            Function
+                      new { x: T => ds }     Object
+
+    Definition  d  =  def a = t              Method definition
+                      type A = T             Type
+
+    Term        t  =  v                      Value
+                      x                      Variable
+                      t1(t2)                 Application
+                      t.a                    Selection
+                      { val x = t1; t2 }     Local definition
+
+    Type        T  =  Any                    Top type
+                      Nothing                Bottom type
+                      x.A                    Selection
+                      (x: T1) => T2          Function
+                      { def a: T }           Method declaration
+                      { type T >: T1 <: T2 } Type declaration
+                      T1 & T2                Intersection
+                      { x => T }             Recursion
+
+The grammar uses several kinds of names:
+
+    x      for (immutable) variables
+    a      for (parameterless) methods
+    A      for types
+
+The full calculus adds to this syntax formal _typing rules_ that assign
+types `T` to terms `t` and formal _evaluation rules_ that describe how a
+program is evaluated. The main difference between the Wadlerfest paper
+and the prior technical report lies in the form of these evaluation
+rules. The technical report uses a "big step" semantics, i.e. it
+defines an interpreter that explains how terms map to
+values. The Wadlerfest paper uses a more common "small step"
+semantics, i.e. it explains evaluation as step-wise rewritings of
+terms. To do this, it needs to restrict terms to be in [A-normal
+form](https://en.wikipedia.org/wiki/A-normal_form), that is, every
+intermediate term is abstracted out in a local definition.
+
+The following _type soundness_ property was shown in each case with a
+mechanized proof:
+
+> If a term `t` has type `T`, and the evaluation of `t` terminates, then
+  the result of the evaluation will be a value `v` of type `T`.
+
+## Difficulties
+
+Formulating the precise soundness theorem and proving it was unexpectedly hard,
+because it uncovered a problem complex that had not been
+studied in depth before. In DOT - as well as in many programming languages -
+you can have conflicting definitions. For instance you might have an abstract
+type declaration in a base class with two conflicting aliases in subclasses:
+
+     trait Base { type A }
+     trait Sub1 extends Base { type A = String }
+     trait Sub2 extends Base { type A = Int }
+
+Now, if you combine `Sub1` and `Sub2` in one class, you get a conflict, since the type `A` is supposed to be equal to both `String` and `Int`. If you do
+not detect the conflict and assume the equalities at face value you
```

Comment:
"Now, if you combine Sub1 and Sub2 in one class" is a crucial step and I guess readers will want to know more precisely what you mean here, e.g. some Scala code like `trait Bad extends Sub1 with Sub2`.